Addition of a cropmark editor to the extension

dodomarg opened the pull request:
In order to mitigate the difficulty of changing the default template I have created a cropmark editor the takes care  of the needed changes to the document.
As TBH I did not fully understand the mechanism of the roland_cutsudio.py and was unable to edit the document to add the required data from the template to the current open document the addition is fully independent to the existing code and simply adds the ability for the user to edit the template once opened in any way they see fit. 
The addition has been tested on a Roland GS2-24 with very good user feedback. 
A behavior I was unable to understand caused the existing 3 cropmark template when sent to CutStudio to show as a four cropmark one, so for the lack of a better solution the editor offers the user 3 or 4 cropmark settings.
Test for the 3 cropmark performance were not carried out due to the issue mentioned above.

dodomarg committed:
Update README.md

dodomarg committed:
Add files via upload

First version of a working editor with presets needs better comments

dodomarg committed:
Update roland_cutstudio_cropmark_template.svg

Changed id of the cropmark layer to be easily removable by the editor

dodomarg committed:
Add files via upload

re upload

dodomarg committed:
Added 4 cutmarks ability to the cropmark editor

A cheap workaround for inability to control the way cutstudio defines the number of cropmarks from Inkscape

dodomarg committed:
Update README.md

dodomarg committed:
Update roland_cropmark_editor.inx

dodomarg committed:
Update README.md

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+
+class PrintingMarks(inkex.EffectExtension):
+
+    # TODO add tool marks for manual alingment
+
+    # TODO impemnent a way to emmbedd the custom XML from template to any file
```
are these comments still up to date?

dodomarg replied:
No, I have removed them

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+            "gr_g": {"top": 30.0, "bottom": 45.0, "left": 10.0, "right": 10.0},
+            "sv_series": {"top": 6.0, "bottom": 30.0, "left": 23.0, "right": 23.0},
+            "sv_8": {"top": 6.0, "bottom": 30.0, "left": 25.0, "right": 24.0},
+        }
```
These values do not match what I found in the Roland documentation. I successfully used the values from there with the GX24, which allows for a slightly larger cutting area:

https://files.rolanddga.com/Files/CutStudioManual/CutStudio/!SSL!/Responsive_HTML5/Hints_and_Tips/crop_position.html

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+                H (float): Height in user units.
+                x (float): X-coordinate for the text.
+                y (float): Y-coordinate for the text.
+                font_size (str): Font size for the text. Default is "9pt".
```
font_size is no argument of the function.
Order of arguments is mixed up in the documentation ("parent" is at the end)

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+            txt = parent.add(inkex.TextElement(**txt_attribs))
+            
+
+            # Construct the cropmark settings string without the extra {{
```
what do you mean with "without the extra {{"?

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+            "id": re.sub(r'\W+', '_', name),  # Replace non-alphanumeric characters with underscores
+            "cx": str(x),       # X position
+            "cy": str(y),       # Y position
+            "r": "5"            # Radius is half of the 10mm diameter
```
Here and in other places you assume that the document unit is millimeters. If I set the unit in the extension dialog e.g. to pt and choose page size A4 then the generated template is very wrong.

dodomarg replied:
Fixed unit conversion issues and the use of the user selected units when non of the custom options are applied.

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+            # Construct the cropmark settings string without the extra {{
+            txt.text = (
+                f'INKSCAPE_CUTSTUDIO_CROPMARK_SETTINGS={{"version":1, '
+                f'"pageW":{pageW}, "pageH":{pageH}, '
+                f'"dx":{dx}, "dy":{dy}, '
+                f'"W":{W}, "H":{H}}}'
```
The values here must always be in millimeters, but your function documentation states they are in user units

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+                y (float): Y-coordinate for the text.
+                font_size (str): Font size for the text. Default is "9pt".
+            """
+            # Convert numeric values to integers to remove decimal points
+            pageW = int(pageW)
+            pageH = int(pageH)
+            dx = int(dx)
+            dy = int(dy)
+            W = int(W)
+            H = int(H)
```
why do you convert to integer here? Custom pagesizes are not guaranteed to be in full millimeters.

dodomarg replied:
By mistake, thanks for pointing out

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+        border_bottom = bbox.bottom - margin_bottom
+
+        # Center lines for cropmarks
+        offset_left = bbox.left + margin_left + 10
+        offset_right = bbox.right - margin_right - 10 
+        offset_top = bbox.top + margin_top + 10
+        offset_bottom = bbox.bottom - margin_bottom - 10
```
Something is wrong with your cropmark computation.
In CutStudio 1.41 I see a difference of about 10 mm. With the old example template everything is correct.

![screenshot](https://github.com/user-attachments/assets/3a2f3fac-0979-4f9d-935e-f5a995f3e83e)

dodomarg replied:
There were some issue. most likely due to the a wrong implemented the 4 crop mark option, seems to have fixed the issue in my fork.

dodomarg replied:
looking at the way they calculate the margins in the [https://files.rolanddga.com/Files/CutStudioManual/CutStudio/!SSL!/Responsive_HTML5/Hints_and_Tips/crop_position.html](https://files.rolanddga.com/Files/CutStudioManual/CutStudio/!SSL!/Responsive_HTML5/Hints_and_Tips/crop_position.html)

It seems that the 3 marks and 4 marks might have to be completely separate implementations as the 3 crop mark version might give 10mm more on account of the manual alignment marks. Currently it mostly means that the margins for the 4 crop marks are narrower than Roland intends and might lead to problems preforming manual alignment.

mgmax commented on README.md:
```diff
@@ -32,14 +32,17 @@ Before using this option, CutStudio must be set up so that normal use works prop
-3. You now have an example file with DIN A4 pagesize and crop marks that should work on a Roland GX-24 (and hopefully many other plotters). Changing the page size and cropmark position is currently not possible.
-4. The following explanation uses separate layers for printing and drawing. You can also do it in a different way.
-5. Draw your print graphics on the layer "Print graphics". Remove the "Remove me" placeholder.
+3. ~~You now have an example file with DIN A4 pagesize and crop marks that should work on a Roland GX-24 (and hopefully many other plotters). Changing the page size and cropmark position is currently not possible.~~
+    You can edit the example file by Extensions -> Roland CutStudio -> Roland Cropmark Editor.
```
```suggestion
3. You now have an example file with DIN A4 pagesize and crop marks that should work on a Roland GX-24 (and hopefully many other plotters). You can change the page size and cropmark position with Extensions -> Roland CutStudio -> Roland Cropmark Editor.
```

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+    # TODO impemnent a way to emmbedd the custom XML from template to any file
+
+    def add_arguments(self, pars):
+        pars.add_argument("--preset", help="Apply crop marks to...", default="gx_24_gs_24")
```
```suggestion
        pars.add_argument("--preset", help="Machine type (or 'custom' to use custom margins)", default="gx_24_gs_24")

```

mgmax commented on roland_cropmark_editor.py:
```diff
@@ -0,0 +1,426 @@
+
+    def add_arguments(self, pars):
+        pars.add_argument("--preset", help="Apply crop marks to...", default="gx_24_gs_24")
+        pars.add_argument("--page_size", help="Apply crop marks to...", default="A1")
```
```suggestion
        pars.add_argument("--page_size", help="Page size (or 'custom' to use custom new_width and new_height, or 'keep' to keep old page size)", default="A1")

```

mgmax reviewed (changes requested):
Thanks for your great suggestion! I tested it and it looks like some bugs still need to be fixed: unit handling, cropmark position calculation and default margins. See the detailed review comments.

mgmax commented:
Did you receive my feedback, do you need any help?

Commit:
1. Fixed the wrong implimentation of crop mark
calculation and the manual alignmen marks drawing
2. Changed order of execution to make more sense
3. Moved all four crop mark actions to the conditional section

dodomarg commented:
Yes, did not have the time to dive into your comment till this week. will see if anything poses a significant challenge and will reach out for help if needed, thanks for offering.

Commit:
Bugfixes and logic improvements
* Seperated the creation of 3 and 4 cropmarks with conditional statments.
* Removed Rounding behavior from width and hight variables in draw_effect .
* Changed variable types to float in add_cropmark_settings_text .
* Updated default margin values based on https://files.rolanddga.com/Files/CutStudioManual/CutStudio/!SSL!/Responsive_HTML5/Hints_and_Tips/crop_position.html
    note - the default values are still a compremise as Roland has a plethra of vaulues for several page sizes and orientations.

dodomarg committed:
Update roland_cropmark_editor.py

update to help text

Co-authored-by: Max Gaukler <mgmax@users.noreply.github.com>

dodomarg committed:
Update README.md

update to readme text

Co-authored-by: Max Gaukler <mgmax@users.noreply.github.com>

dodomarg committed:
Update roland_cropmark_editor.py

update to help text

Co-authored-by: Max Gaukler <mgmax@users.noreply.github.com>

Commit:
Removed un-used variable for inseting the cut area

Commit:
Added:
* unit selection for custom page size
* unit selection for custom margins
* about tab for the cropmark editor

Fixed:
* unit selection used even when user dose not select custom values for page or margins
* comments

dodomarg commented:
Check out the current commit. I think I have addressed all of the concerns, let me know if there is anything else that pops up.

kilzid committed:
Update README.md

dodomarg committed:
Merge pull request #1 from kilzid/patch-1

Update README.md

Commit:
fixed wrong calculation of dy

Commit:
Merge branch 'master' of https://github.com/dodomarg/inkscape-roland-cutstudio

mgmax commented on README.md:
```diff
@@ -64,7 +63,7 @@ Follow the steps above. Once the plugin has completed the export, you'll see a d
 3. Restart Inkscape.
 
-[zip]: https://github.com/mgmax/inkscape-roland-cutstudio/archive/refs/heads/master.zip
+[zip]: https://github.com/dodomarg/inkscape-roland-cutstudio/archive/refs/heads/master.zip
```
Please undo the change:
```suggestion
[zip]: https://github.com/mgmax/inkscape-roland-cutstudio/archive/refs/heads/master.zip
```

mgmax reviewed (changes requested):
Looks good except for the URL in the README that needs to be changed back. Thanks a lot :-)

dodomarg committed:
Update README.md

Co-authored-by: Max Gaukler <mgmax@users.noreply.github.com>

mgmax commented:
Thanks a lot :-)
